### ISSUE TITLE
App.After breaks error checking from App.Before

### ISSUE THREAD

**Issue by chrj** [codegangsta/cli#204]
Hello,

Using `App.After` resets the `err` variable (in `App.Run`) used to capture the error from `App.Before`, so no error is reported.

Here is a sample: 

```go
package main

import (
	"errors"

	"github.com/codegangsta/cli"
)

func main() {

	app := cli.NewApp()

	app.Name = "test"
	app.Usage = "test"
	app.Action = func(c *cli.Context) {}

	app.After = func(c *cli.Context) error { return nil }
	app.Before = func(c *cli.Context) error { return errors.New("test") }

	app.RunAndExitOnError()

}
```

I would expect the error from `App.Before` to get catched by `App.RunAndExitOnError`, which it also does if i remove the `App.After` function.

**Comment by sigmonsays** [comment]
This also seems to be an issue with invalid flags to commands. 

If I setup app.After() any invalid flags are no longer returned to main.

**Comment by sigmonsays** [comment]
It's not quite clear if this is the intended behavior or not. It appears to me that After was tacked on without prior planning due to its signature missing the result. It's also not clear if returning an error in After should terminate the call chain.

I see two possible solutions

1. After should not eat errors
2. After should receive an error from a previous invocation to atleast be allowed to decide how to propgate the error back

**Comment by jszwedko** [comment]
This should be fixed by #230, let me know what you think @sigmonsays. I went with the error aggregate approach to avoid changing the API, but I think your second suggestion is the best if we make a breaking API release.
